api/src/common/utils.rs

```diff
@@ -23,29 +28,90 @@ pub const fn near_data_to_near_token(data: near_api_types::Data<u128>) -> NearTo
 // We default to false as we can't know if an error is critical or not without the types
 // so to keep it safe it's better to retry
 
-pub fn is_critical_blocks_error(err: &RpcError) -> bool {
-    is_critical_json_rpc_error(err, |_| false)
+pub fn to_retry_error<T, E: std::fmt::Debug + Send + Sync>(
+    err: SendRequestError<E>,
+    is_critical_t: impl Fn(&SendRequestError<E>) -> bool,
+) -> RetryResponse<T, SendRequestError<E>> {
+    if is_critical_t(&err) {
+        RetryResponse::Critical(err)
+    } else {
+        RetryResponse::Retry(err)
+    }
 }
 
-pub fn is_critical_validator_error(err: &RpcError) -> bool {
-    is_critical_json_rpc_error(err, |_| false)
+pub fn is_critical_blocks_error(err: &SendRequestError<RpcBlockError>) -> bool {
+    is_critical_json_rpc_error(err, |err| match err {
+        RpcBlockError::UnknownBlock { .. }
+        | RpcBlockError::NotSyncedYet
+        | RpcBlockError::InternalError { .. } => true,
+    })
 }
 
-pub fn is_critical_query_error(rpc_error: &RpcError) -> bool {
-    is_critical_json_rpc_error(rpc_error, |_| false)
+pub fn is_critical_validator_error(err: &SendRequestError<RpcValidatorError>) -> bool {
+    is_critical_json_rpc_error(err, |err| match err {
+        RpcValidatorError::UnknownEpoch
+        | RpcValidatorError::ValidatorInfoUnavailable
+        | RpcValidatorError::InternalError { .. } => true,
+    })
+}
+pub fn is_critical_query_error(err: &SendRequestError<RpcQueryError>) -> bool {
+    is_critical_json_rpc_error(err, |err| match err {
+        RpcQueryError::NoSyncedBlocks
+        | RpcQueryError::UnavailableShard { .. }
+        | RpcQueryError::GarbageCollectedBlock { .. }
+        | RpcQueryError::UnknownBlock { .. }
+        | RpcQueryError::InvalidAccount { .. }
+        | RpcQueryError::UnknownAccount { .. }
+        | RpcQueryError::NoContractCode { .. }
+        | RpcQueryError::TooLargeContractState { .. }
+        | RpcQueryError::UnknownAccessKey { .. }
+        | RpcQueryError::ContractExecutionError { .. }
+        | RpcQueryError::UnknownGasKey { .. }
+        | RpcQueryError::InternalError { .. } => true,
+        // Might be critical, but also might not yet propagated across the network, so we will retry
+        RpcQueryError::NoGlobalContractCode { .. } => false,
+    })
 }
 
-pub fn is_critical_transaction_error(err: &RpcError) -> bool {
-    is_critical_json_rpc_error(err, |_| false)
+pub fn is_critical_transaction_error(err: &SendRequestError<RpcTransactionError>) -> bool {
+    is_critical_json_rpc_error(err, |err| match err {
+        RpcTransactionError::TimeoutError => false,
+        RpcTransactionError::InvalidTransaction { .. }
+        | RpcTransactionError::DoesNotTrackShard
+        | RpcTransactionError::RequestRouted { .. }
```

Comment:
Can RequestRouted, InvliadTransaction::InvalidTxError::{ShardCongested, ShardStuck}, InvalidTransaction::StorageError::StorageInconsistentState be retried? If the answer is not clear, I would retry them